src/Datadog.Trace/Propagation/ContextPropagatorBuilder.cs

```diff
@@ -7,18 +7,25 @@ namespace Datadog.Trace.Propagation
 {
     internal static class ContextPropagatorBuilder
     {
-        private static readonly IReadOnlyDictionary<PropagatorType, Func<ITraceIdConvention, IPropagator>> _propagatorSelector =
+        private static readonly IReadOnlyDictionary<PropagatorType, Func<ITraceIdConvention, IPropagator>> PropagatorSelector =
             new Dictionary<PropagatorType, Func<ITraceIdConvention, IPropagator>>()
             {
-                { PropagatorType.B3, (convention) => new B3SpanContextPropagator(convention) },
-                { PropagatorType.Datadog, (convention) => new DDSpanContextPropagator(convention) },
-                { PropagatorType.Default, (convention) => new DDSpanContextPropagator(convention) },
+                { PropagatorType.W3C, convention => new W3CSpanContextPropagator(convention) },
+                { PropagatorType.B3, convention => new B3SpanContextPropagator(convention) },
+                { PropagatorType.Datadog, convention => new DDSpanContextPropagator(convention) },
+                { PropagatorType.Default, convention => new DDSpanContextPropagator(convention) },
             };
 
         public static IPropagator BuildPropagator(PropagatorType propagator, ITraceIdConvention traceIdConvention)
         {
-            if (_propagatorSelector.TryGetValue(propagator, out Func<ITraceIdConvention, IPropagator> getter))
+            if (PropagatorSelector.TryGetValue(propagator, out Func<ITraceIdConvention, IPropagator> getter))
             {
+                // W3C propagator requires Otel TraceId convention as it's specification clearly states lengths of traceId and spanId values in the header.
+                if (propagator == PropagatorType.W3C && traceIdConvention is not OtelTraceIdConvention)
+                {
+                    throw new NotSupportedException($"'{PropagatorType.W3C}' propagator requires '{ConventionType.OpenTelemetry}' convention to be set");
```

Comment:
@pjanotti @zacharycmontoya @nrcventura 
I am not sure if we should throw an exception and crash the instrumented application in case of improper configuration. As far as I see in all other places there is a fallback to defaults in case of improper value.

Besides, I think we should have some docs to describe the strategies for error handling e.g. when we should log vs crash etc.

Reply:
For me it seems very confusing if the application does not respect the values and starts behaving differently.
1. Either fail fast (+ log the reason of failure)
2. Settings should be consistent for the whole system. So instead of creating default propagator, the setting should be reverting back to the default value (+ log the reason). Otherwise other parts of the system will still see the invalid value but don't know about the actual behavior.  Which means it could lead to weird bugs.

Reply:
I would go with the 1. here, but after giving it a second thought that would make any .NET app crash on the machine which can be disturbing.

Reply:
In general, I am also in favour of 1.
However, maybe there are reasons that we should not go that way. Especially that currently it does not work that way if I follow the code correctly.

Reply:
@dszmigielski I think it's very important point - already forgot that the default behavior is to instruct **every** dotnet application. 

Wondering why the internals of tracer are not protected? If there is a fatal failure it should put the tracer to offline mode (+ log the reason) but leave the original application alive.

Reply:
First, please consult https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/error-handling.md

E.g. in Java most of the (auto-)configuration is done by SDK, which throws an exception and fails to start if configuration is invalid (e.g. user has requested an unknown propagator via env variable)

Reply:
I think we fall under this:
The API or SDK may fail fast and cause the application to fail on initialization, e.g. because of a bad user config or environment, but MUST NOT cause the application to fail later at run time, e.g. due to dynamic config settings received from the Collector.

Reply:
This time it seems ok, but I think using the setting value equals already "later at run time". Using the fail fast model, all of the settings (currently only TracerSettings instance) must be valid max at the end of binding. Everything else is too riski, cause if the setting value is used after initialization, its already "later at run time".

Reply:
> This time it seems ok

TBH now I am not sure. We would need to check what happens if we attach the profiler e.g. to an app that only prints `Hello world` to the console.

Reply:
cc @lucaspimentel 

In this case, it makes sense to bubble up an exception through the Tracer constructor because some combination of configuration was invalid. However, I don't think we account for this currently. We currently use a lazy initializer and get the default `Tracer` instance: https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/blob/57919f03c92c9ccb0917b0fa0c964b85f989dca7/src/Datadog.Trace.ClrProfiler.Managed/Instrumentation.cs#L63 . BTW this is called at startup when the IL rewriting aggressively finds the first JIT-compiled method and makes a call to this: https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/blob/57919f03c92c9ccb0917b0fa0c964b85f989dca7/src/Datadog.Trace.ClrProfiler.Managed.Loader/Startup.cs#L42

Inside `Instrumentation.Initialize` would be a good place to
1. Catch the exception from the constructor
2. Log the exception from the constructor
3. Maybe set `Tracer.Instance = new Tracer(new TracerSettings { TraceEnabled = false } );` so it can't do anymore harm, even though the IL rewriting will still occur?

Reply:
ad. 3
I think it would be better to print also the "log" to the console and close the application. Otherwise, the user may not be aware that the auto-instrumentation is not working.

Reply:
At the moment this will not fail until the first instrumentation attempt is made, so it's no near perfect. Even though I would close this PR and create an issue for fixing this behaviour separately.

Reply:
Yes this issue can be handled separately, and we can discuss what the side effect of bad configuration is. At Datadog, we have a strong opinion that we should absolutely minimize changes to the application's behavior when attaching the tracer, so in our tracer we would not choose to stop the application because of bad configuration.